components/psst/browser/core/psst_rule.cc

```diff
@@ -0,0 +1,135 @@
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace {
+
+// Format:
+// [
+//   {
+//     "include": [
+//       "https://twitter.com/*"
+//     ],
+//     "exclude": [
+//     ],
+//     "version": 1,
+//     "test_script": "twitter/test.js",
+//     "policy_script": "twitter/policy.js"
+//   }, ...
+// ]
+// Note that "test_script" and "policy_script" give paths
+// relative to the component under scripts/
+
+// psst.json keys
+const base::FilePath::CharType kInclude[] = FILE_PATH_LITERAL("include");
+const base::FilePath::CharType kExclude[] = FILE_PATH_LITERAL("exclude");
+const base::FilePath::CharType kTestScript[] = FILE_PATH_LITERAL("test_script");
+const base::FilePath::CharType kPolicyScript[] =
+    FILE_PATH_LITERAL("policy_script");
+const base::FilePath::CharType kVersion[] = FILE_PATH_LITERAL("version");
+
+bool GetURLPatternSetFromValue(const base::Value* value,
+                               extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(value->GetList(), URLPattern::SCHEME_HTTPS,
+                                false, &error);
+  if (!valid) {
+    DVLOG(1) << error;
+  }
+  return valid;
+}
+
+bool GetFilePathFromValue(const base::Value* value, base::FilePath* result) {
+  if (!value->is_string()) {
+    return false;
+  }
+  *result = base::FilePath(value->GetString());
+  return true;
+}
+
+}  // namespace
+
+namespace psst {
+
+PsstRule::PsstRule() = default;
+PsstRule::~PsstRule() = default;
+PsstRule::PsstRule(const PsstRule& other) {
+  include_pattern_set_ = other.include_pattern_set_.Clone();
+  exclude_pattern_set_ = other.exclude_pattern_set_.Clone();
+  test_script_path_ = other.test_script_path_;
+  policy_script_path_ = other.policy_script_path_;
+  version_ = other.version_;
+}
+
+// static
+void PsstRule::RegisterJSONConverter(
+    base::JSONValueConverter<PsstRule>* converter) {
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kInclude, &PsstRule::include_pattern_set_, GetURLPatternSetFromValue);
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kExclude, &PsstRule::exclude_pattern_set_, GetURLPatternSetFromValue);
+  converter->RegisterCustomValueField<base::FilePath>(
+      kTestScript, &PsstRule::test_script_path_, GetFilePathFromValue);
+  converter->RegisterCustomValueField<base::FilePath>(
+      kPolicyScript, &PsstRule::policy_script_path_, GetFilePathFromValue);
+  converter->RegisterIntField(kVersion, &PsstRule::version_);
+}
+
+// static
+absl::optional<std::vector<PsstRule>> PsstRule::ParseRules(
+    const std::string& contents) {
+  if (contents.empty()) {
+    return absl::nullopt;
+  }
+  absl::optional<base::Value> root = base::JSONReader::Read(contents);
+  if (!root) {
+    return absl::nullopt;
+  }
+  std::vector<PsstRule> rules;
+  base::JSONValueConverter<PsstRule> converter;
+  for (base::Value& it : root->GetList()) {
```

Comment:
This loop here, for instance, wouldn't be necessary with an IDL, with all the conversion logic being tool generated.

Reply:
This is following an old pattern for `*Rule` that we really should fix and templatize. I think this would be good for a follow-up to do this one correctly so we don't keep propagating this